### PR TITLE
Couch core

### PR DESCRIPTION
**Pull request by tduehr**
Breaking up #3357 for auditability.

This PR contains the core changes to support the creation of the remaining features.

### CODE REVIEW

File: support/cas-server-support-couchdb-core/src/main/resources/org/apereo/cas/couchdb/CAS_delete_record.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+function(doc, req){
```
Comment (mmoayyed) [inline support/cas-server-support-couchdb-core/src/main/resources/org/apereo/cas/couchdb/CAS_delete_record.js]: Should this be removed?

File: docs/cas-server-documentation/installation/Ticket-Registry-Replication-Encryption.md [hunk 2]
```diff
@@ -10,6 +10,7 @@ by encrypting and signing tickets:
 * [Hazelcast](Hazelcast-Ticket-Registry.html)
 * [Ehcache](Ehcache-Ticket-Registry.html)
 * [Ignite](Ignite-Ticket-Registry.html)
+* [CouchDb](CouchDb-Ticket-Registry.html)
```
Comment (mmoayyed) [inline docs/cas-server-documentation/installation/Ticket-Registry-Replication-Encryption.md]: Also list here? https://apereo.github.io/cas/development/installation/Configuring-Ticketing-Components.html

File: support/cas-server-support-couchdb-core/src/main/java/org/apereo/cas/couchdb/core/CouchDbConnectorFactory.java [hunk 3]
```diff
@@ -23,19 +26,69 @@
...
 @RequiredArgsConstructor
+@Slf4j
 public class CouchDbConnectorFactory {
 
-    private final AbstractCouchDbProperties couchDbProperties;
+    private final BaseCouchDbProperties couchDbProperties;
+    private final ObjectMapperFactory objectMapperFactory;
     private CouchDbConnector couchDbConnector;
     private CouchDbInstance couchDbInstance;
+    private HttpClient httpClient;
```
Comment (mmoayyed) [inline support/cas-server-support-couchdb-core/src/main/java/org/apereo/cas/couchdb/core/CouchDbConnectorFactory.java]: It really looks like these objects are treated as singletons? If so, why not make them be singletons by following that pattern?

File: support/cas-server-support-couchdb-core/src/main/java/org/apereo/cas/couchdb/core/CouchDbConnectorFactory.java [hunk 4]
```diff
@@ -50,17 +103,24 @@ public CouchDbConnector create() {
...
-            builder.username(couchDbProperties.getUsername());
+        if (StringUtils.isNotBlank(couchDbProperties.getProxyHost())) {
+            builder.proxy(couchDbProperties.getProxyHost());
+
+            if (couchDbProperties.getProxyPort() > 0) {
+                builder.proxyPort(couchDbProperties.getProxyPort());
+                LOGGER.warn("CouchDb proxy settings enabled [{}:{}]", couchDbProperties.getProxyHost(), couchDbProperties.getProxyPort());
+            } else {
+                LOGGER.error("Proxy port not set for host [{}] clearing proxy host.", couchDbProperties.getProxyHost());
+                builder.proxy(null);
```
Comment (mmoayyed) [inline support/cas-server-support-couchdb-core/src/main/java/org/apereo/cas/couchdb/core/CouchDbConnectorFactory.java]: Are the above log statements really, errors and warnings? Seems like debug and info should suffice just fine.

File: support/cas-server-support-couchdb-ticket-registry/src/main/java/org/apereo/cas/config/CouchDbTicketRegistryConfiguration.java [hunk 5]
```diff
@@ -30,10 +31,13 @@
     @Autowired
     private CasConfigurationProperties casProperties;
 
+    @Autowired
+    private ObjectMapperFactory objectMapperFactory;
```
Comment (mmoayyed) [inline support/cas-server-support-couchdb-ticket-registry/src/main/java/org/apereo/cas/config/CouchDbTicketRegistryConfiguration.java]: Probably should use a qualifier here to make overrides easier.

File: support/cas-server-support-couchdb-core/src/main/resources/org/apereo/cas/couchdb/CAS_delete_record.js (hunk 1 above)
Comment (tduehr) [reply support/cas-server-support-couchdb-core/src/main/resources/org/apereo/cas/couchdb/CAS_delete_record.js]: Probably... it's not used by anything but i left it there as a starter for other stuff...

File: support/cas-server-support-couchdb-core/src/main/java/org/apereo/cas/couchdb/core/CouchDbConnectorFactory.java (hunk 3 above)
Comment (tduehr) [reply support/cas-server-support-couchdb-core/src/main/java/org/apereo/cas/couchdb/core/CouchDbConnectorFactory.java]: Not my classes... They wouldn't be singletons if a user put different databases in different couchdb servers. I might wind up doing that to separate the higher traffic ones from the lower...

File: support/cas-server-support-couchdb-core/src/main/java/org/apereo/cas/couchdb/core/CouchDbConnectorFactory.java (hunk 4 above)
Comment (tduehr) [reply support/cas-server-support-couchdb-core/src/main/java/org/apereo/cas/couchdb/core/CouchDbConnectorFactory.java]: Dont

File: support/cas-server-support-couchdb-ticket-registry/src/main/java/org/apereo/cas/config/CouchDbTicketRegistryConfiguration.java (hunk 5 above)
Comment (tduehr) [reply support/cas-server-support-couchdb-ticket-registry/src/main/java/org/apereo/cas/config/CouchDbTicketRegistryConfiguration.java]: Done

File: docs/cas-server-documentation/installation/Ticket-Registry-Replication-Encryption.md (hunk 2 above)
Comment (tduehr) [reply docs/cas-server-documentation/installation/Ticket-Registry-Replication-Encryption.md]: It is, in the NoSQL stuff.

File: support/cas-server-support-couchdb-core/src/main/resources/org/apereo/cas/couchdb/CAS_delete_record.js (hunk 1 above)
Comment (tduehr) [reply support/cas-server-support-couchdb-core/src/main/resources/org/apereo/cas/couchdb/CAS_delete_record.js]: Done

File: style/checkstyle-suppressions.xml [hunk 6]
```diff
@@ -8,7 +8,7 @@
     <suppress checks="IllegalCatch" files="(.*Aspect)\.java"/>
     <suppress checks="JavadocStyleCheck" files="(.*Tests*|Mock.*|Test.*)\.java"/>
     <suppress checks="Indentation" files="(.*Tests*|Mock.*|Test.*)\.java"/>
-    <suppress checks="JavadocMethod" files="(.*Configuration|.*Tests*|Mock.*|Test.*)\.java"/>
+    <suppress checks="JavadocMethod" files="(.*(Configuration|CouchDbRepository)|.*Tests*|Mock.*|Test.*)\.java"/>
```
Comment (mmoayyed) [inline style/checkstyle-suppressions.xml]: This should be removed.

File: support/cas-server-support-consent-couchdb/src/test/resources/log4j2.xml [hunk 7]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<Configuration shutdownHook="disable">
+    <Appenders>
+        <Console name="console" target="SYSTEM_OUT">
+            <PatternLayout pattern="%d %p [%c] - &lt;%m&gt;%n"/>
+        </Console>
+        <RollingFile name="file" fileName="build/gauthcouchdb.log" append="true"
```
Comment (mmoayyed) [inline support/cas-server-support-consent-couchdb/src/test/resources/log4j2.xml]: Change log file name?

File: support/cas-server-support-couchdb-core/src/main/java/org/apereo/cas/config/CasCouchDbCoreConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+/**
+ * This is {@link CasCouchDbCoreConfiguration}, defines certain beans via configuration
+ * while delegating some to Spring namespaces inside the context config file.
+ *
+ * @author Timur Duehr
+ * @since 6.0.0
+ */
+@Configuration("casCouchDbCoreConfiguration")
+@EnableConfigurationProperties(CasConfigurationProperties.class)
+public class CasCouchDbCoreConfiguration {
```
Comment (mmoayyed) [inline support/cas-server-support-couchdb-core/src/main/java/org/apereo/cas/config/CasCouchDbCoreConfiguration.java]: I dont see how this file is mapped in spring.factories. Is that done here?

File: support/cas-server-support-couchdb-core/src/test/resources/log4j2.xml [hunk 9]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<Configuration shutdownHook="disable">
+    <Appenders>
+        <Console name="console" target="SYSTEM_OUT">
+            <PatternLayout pattern="%d %p [%c] - &lt;%m&gt;%n"/>
+        </Console>
+        <RollingFile name="file" fileName="build/gauthcouchdb.log" append="true"
```
Comment (mmoayyed) [inline support/cas-server-support-couchdb-core/src/test/resources/log4j2.xml]: Likely copy/paste error, for the file name.

File: support/cas-server-support-couchdb-core/src/main/java/org/apereo/cas/couchdb/core/CouchDbConnectorFactory.java (hunk 3 above)
Comment (tduehr) [reply support/cas-server-support-couchdb-core/src/main/java/org/apereo/cas/couchdb/core/CouchDbConnectorFactory.java]: lazy initialization, it's so the create methods can be called more than once... probably should rename them...

File: style/checkstyle-suppressions.xml (hunk 6 above)
Comment (tduehr) [reply style/checkstyle-suppressions.xml]: bummer...

File: support/cas-server-support-couchdb-core/src/main/java/org/apereo/cas/config/CasCouchDbCoreConfiguration.java (hunk 8 above)
Comment (tduehr) [reply support/cas-server-support-couchdb-core/src/main/java/org/apereo/cas/config/CasCouchDbCoreConfiguration.java]: Done...

File: support/cas-server-support-couchdb-core/src/test/resources/log4j2.xml (hunk 9 above)
Comment (tduehr) [reply support/cas-server-support-couchdb-core/src/test/resources/log4j2.xml]: Done

File: support/cas-server-support-couchdb-core/src/main/java/org/apereo/cas/couchdb/core/CouchDbConnectorFactory.java (hunk 3 above)
Comment (tduehr) [reply support/cas-server-support-couchdb-core/src/main/java/org/apereo/cas/couchdb/core/CouchDbConnectorFactory.java]: Done

File: core/cas-server-core-util-api/src/main/java/org/apereo/cas/util/CollectionUtils.java [hunk 10]
```diff
@@ -97,6 +97,44 @@
         return c;
     }
 
+    /**
+     * Convert the object given into a {@link List} instead.
+     *
+     * @param obj the object to convert into a collection
+     * @return The collection instance containing the object provided
+     */
+    public static List<Object> toList(final Object obj) {
```
[the remark this answers is not in the document]
Comment (tduehr) [reply core/cas-server-core-util-api/src/main/java/org/apereo/cas/util/CollectionUtils.java]: Done... seems to have already been removed from usage...

File: core/cas-server-core-util-api/src/main/java/org/apereo/cas/util/CollectionUtils.java [hunk 11]
```diff
@@ -97,6 +97,44 @@
...
+    }
+
+    /**
+     * Cast object to {@link List} with appropriate type.
+     * @param obj Object to cas.
+     * @param klass Class to cast elements
+     * @param <T> type for casting
+     * @return List with elements cast to type
+     */
+    public static <T> List<T> toList(final Object obj, final Class<T> klass) {
```
[the remark this answers is not in the document]
Comment (tduehr) [reply core/cas-server-core-util-api/src/main/java/org/apereo/cas/util/CollectionUtils.java]: see above...

**Pull request merged**